### PR TITLE
Add support for payment method updates.

### LINKED ISSUE #22: Add support for payment method updates
From customer:

> We use Restrict Content Pro with Pronamic for the subscriptions. If the user's card expires, we lose the user.
> 
> Can we use the `rcp_update_card` shortcode for payment method updates
> - https://restrictcontentpro.com/knowledgebase/rcp_update_card/
> 
> Or is that unsupported with Pronamic and should we use a different way?

We can hook into `rcp_membership_can_update_billing_card` and use the subscription mandate selection URL as action URL on the RCP account page. The `rcp_update_card` shortcode can only be used to display a form like this:

file-hJ90kgI3KT

### PR DESCRIPTION
**Pull request by rvdsteege**
Fix #22.

### CODE REVIEW

File: src/Extension.php [hunk 1]
```diff
@@ -599,6 +617,125 @@ public function rcp_membership_payment_profile_cancelled( $success, $gateway, $g
...
+
+		foreach ( $payments as $payment ) {
+			if ( 'subscription_payment_method_change' !== $payment->get_source() ) {
+				continue;
+			}
+
+			$payment_methods[] = $payment->get_payment_method();
+
+			break;
+		}
```
Comment (remcotolsma) [inline src/Extension.php:709]: This is not scalable, in theory hundreds of payments can be linked to a subscription.

File: src/Extension.php [hunk 2]
```diff
@@ -599,6 +617,125 @@ public function rcp_membership_payment_profile_cancelled( $success, $gateway, $g
...
+			}
+
+			$new_link = new WP_HTML_Tag_Processor( $link );
+
+			if ( $new_link->next_tag( 'a' ) ) {
+				$new_link->set_attribute( 'href', $subscription->get_mandate_selection_url() );
+			}
+
+			$actions = \str_replace( $link, $new_link, $actions );
+		}
```
Comment (remcotolsma) [inline src/Extension.php:667]: I think the use of the `WP_HTML_Tag_Processor` indicates that it is not intended this way.

File: src/Extension.php [hunk 3]
```diff
@@ -85,6 +87,20 @@ public function setup() {
 		\add_action( 'plugins_loaded', [ $this, 'plugins_loaded' ], 5 );
 
 		\add_action( 'rcp_after_membership_admin_update', [ $this, 'rcp_after_membership_admin_update' ] );
+
+		/*
+		 * Filter if billing can be updated for membership.
+		 *
+		 * @link https://gitlab.com/pronamic-plugins/restrict-content-pro/-/blob/3.4.4/includes/memberships/class-rcp-membership.php#L3231-3240
+		 */
+		\add_filter( 'rcp_membership_can_update_billing_card', [ $this, 'rcp_membership_can_update_billing' ], 10, 2 );
```
Comment (remcotolsma) [inline src/Extension.php:96]: This filter is intended to indicate whether a credit card can be updated, if I understand correctly this has now been used more widely. It provides a link to the "Update Billing Card" page with the `[rcp_update_card]` shortcode. This shortcode really shows fields specifically for a credit card:

- Name on Card
- Card Number
- Card CVC
- Card ZIP or Postal Code
- Expiration (MM/YYYY)

https://github.com/stellarwp/restrict-content/blob/386b37e8fe58854c3816fd5c3de451a896737d15/core/templates/card-update-form-fields.php#L40

Using (or abusing) this filter to also update SEPA direct debit mandates does not seem to me to be sufficiently future-proof?

**Review by remcotolsma — CHANGES_REQUESTED** [review]

File: src/Extension.php (hunk 2 above)
Comment (rvdsteege) [reply src/Extension.php]: Changed in https://github.com/pronamic/wp-pronamic-pay-restrict-content-pro/pull/23/commits/0c46dfe43fc27d2df6800f287e3141988d38fba7.

File: src/Extension.php (hunk 3 above)
Comment (rvdsteege) [reply src/Extension.php]: Removed hooking into `rcp_membership_can_update_billing_card` filter in https://github.com/pronamic/wp-pronamic-pay-restrict-content-pro/pull/23/commits/0c46dfe43fc27d2df6800f287e3141988d38fba7.

File: src/Extension.php [hunk 4]
```diff
@@ -599,6 +610,115 @@ public function rcp_membership_payment_profile_cancelled( $success, $gateway, $g
...
+					'value' => 'subscription_payment_method_change',
+				],
+				[
+					'key'   => '_pronamic_payment_subscription_id',
+					'value' => $subscription->get_id(),
+				],
+			],
+		];
+
+		$payment = \get_pronamic_payment_by_meta( '', '', $args );
```
[the remark this answers is not in the document]
Comment (rvdsteege) [reply src/Extension.php:688]: A subscription payment method change with e.g. iDEAL results in the subscription payment method being updated to Direct Debit.

[the remark this answers is not in the document]
Comment (remcotolsma) [reply src/Extension.php:688]: Is that a problem?

[the remark this answers is not in the document]
Comment (rvdsteege) [reply src/Extension.php]: Discussed at office. I thought it was a problem to update the RCP membership gateway to a gateway which was not enabled, but it seems that is no issue. Changed in https://github.com/pronamic/wp-pronamic-pay-restrict-content-pro/pull/23/commits/c40b2a308c94e179a8d255c7d0d10fc402b07669.

**Review by remcotolsma — APPROVED** [review]

**Pull request merged**